### ISSUE TITLE
`\prg_generate_conditional_variant:Nnn` fails to produce new conditions for existing variants and new variants for conditions that do not exist in the base variant of a conditional function

### ISSUE THREAD

**Issue by Witiko** [latex3/latex3#1189]
With regular expl3 functions, the author only has to define the base form of their function using the argument types that are the most convenient for them:

``` tex
\cs_new:Nn
  \mymodule_greet:n
  { Hello~#1! }
```

The caller can then generate and use any variant of the function that is most convenient for them and expl3 will automatically coerce the arguments:

``` tex
\tl_new:N
  \l_mymodule_name_tl
\tl_set:Nn
  \l_mymodule_name_tl
  { world }
\cs_generate_variant:Nn
  \mymodule_greet:n
  { v }
\mymodule_greet:v
  { l_mymodule_name_tl }
```

> Output: Hello world!

This makes life easier for everyone and is well-worth the extra boilerplate introduced by `\cs_generate_variant:Nn`.

The same does not seem to be true for conditional functions, where the author can make life difficult for the caller if they are not careful. Assume that the author has created the `\mymodule_is_prime:n` conditional function with only the `TF` condition:

``` tex
\prg_new_conditional:Nnn
  \mymodule_is_prime:n
  { TF }
  {
    \int_compare:nTF { #1 < 2 }
      { \prg_return_false: }
      {
        \bool_set_false:N \l_tmpa_bool
        \int_step_inline:nnnn { 2 } { 1 } { #1 - 1 } {
          \int_compare:nT { \int_mod:nn { #1 } { ##1 } = 0 } {
            \bool_set_true:N \l_tmpa_bool
          }
        }
        \bool_if:NTF
          \l_tmpa_bool
          { \prg_return_false: }
          { \prg_return_true: }
    }
}
```

If the caller uses just `\mymodule_is_prime:nTF`, all is well:

``` tex
Is~it~prime?!~
\mymodule_is_prime:nTF
  { 41 }
  { Yes,~it~is! }
  { No,~it~is~not. }
```

> Output: Is it prime?! Yes, it is!

However, if the caller wants to add the `T` condition to the base variant `\mymodule_is_prime:n`, an infinite loop is produced:

``` tex
\prg_generate_conditional_variant:Nnn
  \mymodule_is_prime:n
  { n }
  { T }
Is~it~prime?!~
\mymodule_is_prime:nT
  { 41 }
  { Yes,~it~is! }
```

> Expected output: Is it prime?! Yes, it is!
> Actual output: None, doesn't halt.

Furthermore, when the caller adds the `x` variant for a condition `F` that is not defined for the base variant `\mymodule_is_prime:n`, expanding the generated conditional function `\mymodule_is_prime:xF` has no effect:

``` tex
\prg_generate_conditional_variant:Nnn
  \mymodule_is_prime:n
  { x }
  { F }
Is~it~prime?!~
\mymodule_is_prime:xF
  { 41 }
  { No,~it~is~not. }
```

> Expected output: None
> Actual output: Is it prime?! 41No, it is not.

Unless I am using `\prg_generate_conditional_variant:Nnn` incorrectly, it seems that the author can make life difficult for the caller unless they define all conditions (`TF, T, F, p`) for their conditional functions, because it is not straightforward to define additional conditions later. It seems useful to have this limitation of `\prg_generate_conditional_variant:Nnn` et al. either clearly documented in `interfaces.pdf` or fixed.

**Comment by PhelypeOleinik** [comment]
On the one hand, I am 71,34% sure that somewhere in the expl3 documentation there is a paragraph saying something along the lines of "public conditionals should have all the branching versions (and the predicate version, if appropriate) defined", which is a reasonable requirement, I think (there is really no reason _not to_ provide them.

On the other hand, I see no reason why `\prg_generate_conditional_variant:Nnn` couldn't be extended to generate those missing versions (currently it probably just assume they exist).

**Comment by Skillmon** [comment]
This will have `\prg_generate_conditional_variant:Nnn` have to guess from which base to derive the alternative branching version. So your code will have to check whether the "correct" base-variant exists, and if it doesn't create it from any of the others (if one of them exists).

Something like (in some adhoc pseudo-code):

```none
if not exists(base) then
  if exists(base-TF) then
    create_branch_from_TF()
  else if exists(base-T) then
    create_branch_from_T()
  else if exists(base-F) then
    create_branch_from_F()
  else if exists(base-p) then
    create_branch_from_p()
  else
    error(undefined conditional)
    return
  endif
endif
generate_variant()
return
```

**Comment by Witiko** [comment]
@Skillmon Should not be a significant slowdown, since we'd do that only once for a given condition of a conditional function in a single compile cycle, especially if you generate the base variant as well. For example, generating `{ x } { T }` would generate `{ n } { T }` as well, so that we don't have to do the guesswork again when we are generating `{ v } { T }`

However, if the documentation already says that public conditional functions should always define all conditions as @PhelypeOleinik suggested, then adding code to handle this seems as a difficult-to-justify development and maintenance cost.

**Comment by PhelypeOleinik** [comment]
And (continuing @Skillmon's comment) it assumes the conditional was created with `\prg_new_conditional:Npnn`, rather than hand-crafted with `\cs_new:Npn` which is (not really supported but) very much possible if the author of the code intended some optimisation. For example, `\sys_if_engine_pdftex:TF` is equal to `\use_i:nn` (or `\use_ii:nn`) so it's not easy to generate a different branching version of something like that.

**Comment by PhelypeOleinik** [comment]
> However, if the documentation already says that public conditional functions should always define all conditions as @PhelypeOleinik suggested

It's been a while since I've seen that, so I'm not sure if that note really exists or if I'm hallucinating :)

**Comment by Witiko** [comment]
I am not sure `\sys_if_engine_pdftex:TF` is a good example:
``` tex
\cs_new:Nn
  \sys_if_engine_pdftex:T
  {
    \sys_if_engine_pdftex:TF
      { #1 }
      { }
  }
```
However, there does not seem to be a straightforward way to convert `T` and `F` to `TF` if you treat the functions as black boxes.

**Comment by Witiko** [comment]
I am in favor of just having this documented, not necessarily "fixed" (assuming this is a bug).

**Comment by FrankMittelbach** [comment]
I think defining conditionals is special enough that it is sufficient (and advisable) to just properly document that for publicly available conditionals all variants need to be defined up front.

**Comment by Skillmon** [comment]
@PhelypeOleinik Similar to what @Witiko suggests, the different branching versions can just be generated with `\use_i:nn` and friends appended to the found branch-version. Since `\prg_new_conditional:Nnn` might add optimisations you'd have to cater for two different cases if you'd always assume it was used to create the conditionals.

**Comment by Skillmon** [comment]
> However, there does not seem to be a straightforward way to convert e.g. the variants `T` and `F` to `TF` if you treat the functions as black boxes.

Assuming the argument specification is correct and doesn't use `w` you can simply add `\use_ii:nnn \use_ii:nn` or `\use_iii:nnn \use_i:nn` after the `T` or `F` variant to get `TF`. The only ill-mannered thing would be some conditional `\my_if_foo:wT`.

**Comment by alceu-frigeri** [comment]
Not sure if it is just the case of being more explicit (in the manual) about conditional variants (and the need to declare the full set).

But, I stumble on that:
```latex
 \documentclass{article}
\RequirePackage{expl3}
\ExplSyntaxOn
\tl_new:N \l__starray_prefix_tl

\prg_new_protected_conditional:Npnn \__starray_if_free:n #1 {T, F, TF}
  {
    \prop_if_exist:cTF {#1}
      { \prg_return_false: }
      { \prg_return_true: }
  }
\prg_generate_conditional_variant:Nnn \__starray_if_free:n {e} {p, T, F, TF}

\prg_new_protected_conditional:Npnn \starray_new:n #1 { T, F, TF}
  {
    \__starray_if_free:nTF {\l__starray_prefix_tl #1 _def_prop}
      {  \prg_return_true: }
      { \prg_return_false: }
  }
\prg_generate_conditional_variant:Nnn \starray_new:n {e} {p , T, F, TF}

\begin{document}
 this ~  is ~ just ~ a  ~ test
 
 \bool_if:nTF { \starray_new_p:e {myfoo} }
   { \message { TRUE } }
   { \message { FALSE } }
\end{document}
\ExplSyntaxOff
```

which obviously fails, but only at the ```\bool_if``` 
```\prg_generate_conditional_variant``` raises no warning that the predicate wasn't present (or that the command isn't expandable...)

**Comment by PhelypeOleinik** [comment]
@alceu-frigeri Hm... that seems like a geniune bug.  `\prg_generate_conditional_variant:Nnn` shouldn't generate a predicate version of a protected conditional.

**Comment by muzimuzhi** [comment]
I think https://github.com/latex3/latex3/issues/1189#issuecomment-1551314802 should be reported in a separate issue.

```diff
diff --git a/l3kernel/l3expan.dtx b/l3kernel/l3expan.dtx
index 28ac1c8..23b0983 100644
--- a/l3kernel/l3expan.dtx
+++ b/l3kernel/l3expan.dtx
@@ -2247,7 +2247,14 @@
     \@@_generate_variant:w #2 \s_@@_mark {#3} {#4} {#5}
   }
 \cs_new_protected:Npn \@@_generate_variant_p_form:nnn #1#2
-  { \cs_generate_variant:cn { #1 _p : #2 } }
+  {
+    \cs_if_exist:cTF {#1 _p : #2}
+      { \cs_generate_variant:cn { #1 _p : #2 } }
+      {
+        % some error
+        \use_none:nn
+      }
+  }
 \cs_new_protected:Npn \@@_generate_variant_T_form:nnn #1#2
   { \cs_generate_variant:cn { #1 : #2 T } }
 \cs_new_protected:Npn \@@_generate_variant_F_form:nnn #1#2
```

**Comment by PhelypeOleinik** [comment]
> @PhelypeOleinik Are you happy to take this one?

Sure

**Comment by josephwright** [comment]
Likely solution here is to rethink how to generate T/F/TF/p versions in the first place

**Comment by josephwright** [comment]
Auto-coercision would look something like
```latex
\ExplSyntaxOn
\cs_gset_protected:Npn \__cs_generate_variant:nnNnn #1#2#3#4#5
  {
    \if_meaning:w \c_false_bool #3
      \msg_error:nne { kernel } { missing-colon }
        { \token_to_str:c {#1} }
      \__cs_use_i_delimit_by_s_stop:nw
    \fi:
    \__cs_generate_variant_check:nn {#1} {#2}
    \exp_after:wN \__cs_generate_variant:w
    \tl_to_str:n {#5} , \scan_stop: , \q__cs_recursion_stop
    \__cs_use_none_delimit_by_s_stop:w \s__cs_mark {#1} {#2} {#4} \s__cs_stop
  }
\cs_new_protected:Npn \__cs_generate_variant_check:nn #1#2
  {
    \cs_if_exist:cF { #1 : #2 TF }
      {
        \cs_if_exist:cF { #1 : #2 T }
          { \cs_if_exist:cF { #1 : #2 F }
            { \cs_if_exist:cF { #1 _p : #2 }
              {
                \ERROR
                \__cs_use_none_delimit_by_s_stop:w
              }
            }  
         }
    }
  }
\cs_gset_protected:Npn \__cs_generate_variant_p_form:nnn #1#2
  { \cs_generate_variant:cn { #1 _p : #2 } }
\cs_gset_protected:Npn \__cs_generate_variant_T_form:nnn #1#2
  {
    \__cs_generate_variant_ensure:nnn {#1} {#2} { T }
    \cs_generate_variant:cn { #1 : #2 T }
  }
\cs_gset_protected:Npn \__cs_generate_variant_F_form:nnn #1#2
  {
    \__cs_generate_variant_ensure:nnn {#1} {#2} { F }
    \cs_generate_variant:cn { #1 : #2 F }
  }
\cs_gset_protected:Npn \__cs_generate_variant_TF_form:nnn #1#2
  {
    \__cs_generate_variant_ensure:nnn {#1} {#2} { TF }
    \cs_generate_variant:cn { #1 : #2 TF }
  }
\cs_new_protected:Npn \__cs_generate_variant_ensure:nnn #1#2#3
  {
    \cs_if_exist:cF { #1 : #2 #3 }
      {
        \cs_if_exist:cTF { #1 : #2 TF }
          {
            \__cs_generate_variant:c { #1 : #2 TF }
            \exp_args:Nc \__cs_tmp:w { #1 : #2 #3 }
              {
                \exp_not:c { #1 : #2 TF }
                \use:c { __cs_generate_variant_TF_ #3 : }
              }
          }
          {
            \cs_if_exist:cTF { #1 : #2 T }
              {
                \__cs_generate_variant:c { #1 : #2 T }
                \exp_args:Nc \__cs_tmp:w { #1 : #2 #3 }
                  {
                    \exp_not:c { #1 : #2 T }
                    \use:c { __cs_generate_variant_T_ #3 : }
                  }
              }
              {
                \cs_if_exist:cTF { #1 : #2 F }
                  {
                    \__cs_generate_variant:c { #1 : #2 F }
                    \exp_args:Nc \__cs_tmp:w { #1 : #2 #3 }
                      {
                        \exp_not:c { #1 : #2 F }
                        \use:c { __cs_generate_variant_F_ #3 : }
                      }   
                  }
                  {
                    \cs_new:cpe { #1 : #2 #3 }
                      {
                        \exp_not:N \bool_if:NTF \exp_not:c { #1 _p: #2 #3 }
                        \use:c { __cs_generate_variant_TF_ #3 : }
                      }
                  }
          }
      }
  }
\cs_new:Npn \__cs_generate_variant_T_TF:
  { \exp_not:N \use_i:nn \exp_not:N \use_none:n }
\cs_new:Npn \__cs_generate_variant_T_F:
  { \exp_not:N \use_none:nn \exp_not:N \use:n }
\cs_new:Npn \__cs_generate_variant_F_T:
  { \exp_not:N \use_ii:nn \exp_not:N \use:n }
\cs_new:Npn \__cs_generate_variant_F_TF:
  { \exp_not:N \use_i:nnn \exp_not:N \use_ii:nn }
\cs_new:Npn \__cs_generate_variant_TF_T:
  { \exp_not:N \use:n \exp_not:N \use_none:n }
\cs_new:Npn \__cs_generate_variant_TF_F:
  { \exp_not:N \use_none:n \exp_not:N \use:n }
\cs_generate_variant:Nn \__cs_generate_variant:N { c }
\prg_new_conditional:Npnn \foo:N #1 { TF } { \prg_return_true: }

\prg_generate_conditional_variant:Nnn \foo:N { c }  { F }
```
Do we want a PR doing this?

**Comment by josephwright** [comment]
(Generation of `p` form outstanding of course.)

**Comment by josephwright** [comment]
Do we want an alternative PR for the above? I'll definitely add some notes to the docs whatever else happens
